### PR TITLE
👷 implement merge into staging job

### PR DESCRIPTION
**Pull request by amortemousque**
## Motivation

We want to setup a staging workflow similar to web-ui: instead of deploying main on staging, a new staging-xx branch is created every week. This branch is updated on every main commits and will be deployed on staging. PRs can be merged to this branch to allow testing them early.

This PR implement the merge into staging job, which merge main into staging on each main commit.

## Changes

- Create merge into staging script
- Update gitlab-ci

## Testing

Locally, Gitlab

---

I have gone over the [contributing](https://github.com/DataDog/browser-sdk/blob/main/CONTRIBUTING.md) documentation.

### CODE REVIEW

File: scripts/utils.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+function printLog(...params) {
+  const greenColor = '\x1b[32;1m'
+  console.log(greenColor, ...params, resetColor)
+}
+
+module.exports = {
+  initGitConfig,
+  executeCommand,
+  printError,
+  printLog,
```
Comment (bcaudan) [inline scripts/utils.js:59]: Could we use some of them in existing scripts?

Comment (amortemousque) [reply scripts/utils.js:59]: Yes, I was thinking to do it in another PR.

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -151,6 +158,7 @@ unit-bs:
     refs:
       - main
       - tags
+      - /^staging-[0-9]*$/
```
Comment (bcaudan) [inline .gitlab-ci.yml:161]:
```suggestion
      - /^staging-[0-9]+$/
```
nit

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -4,6 +4,8 @@ variables:
   CURRENT_CI_IMAGE: 25
   BUILD_STABLE_REGISTRY: '486234852809.dkr.ecr.us-east-1.amazonaws.com'
   CI_IMAGE: '$BUILD_STABLE_REGISTRY/ci/$APP:$CURRENT_CI_IMAGE'
+  GIT_REPOSITORY: 'git@github.com:DataDog/browser-sdk.git'
```
Comment (bcaudan) [inline .gitlab-ci.yml:7]: Could `CI_REPOSITORY_URL` predefined env variable work? or is it the gitlab url?

Comment (amortemousque) [reply .gitlab-ci.yml:7]: Sadly it references the gitlab url

Comment (bcaudan) [reply .gitlab-ci.yml:7]: kinda make sense

File: scripts/utils.js (hunk 1 above)
Comment (bcaudan) [reply scripts/utils.js:59]: 👌

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -357,3 +370,18 @@ staging-reset-scheduled-failure:
...
+  only:
+    refs:
+      - main
+  except:
+    refs:
+      - schedules
+  before_script:
+    - eval $(ssh-agent -s)
+  script:
+    - node --no-warnings scripts/staging-ci/merge-into-staging.js
```
Comment (BenoitZugmeyer) [inline .gitlab-ci.yml:387]: Why do we need the `--no-warnings` flag?

**Review by bcaudan — APPROVED** [review]
LGTM

File: scripts/staging-ci/merge-into-staging.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  await initGitConfig(REPOSITORY)
+  await executeCommand(`git fetch --no-tags origin ${CURRENT_STAGING_BRANCH}`)
+  await executeCommand(`git checkout ${CURRENT_STAGING_BRANCH} -f`)
+  await executeCommand(`git pull origin ${CURRENT_STAGING_BRANCH}`)
+
+  printLog(`Merging ${CI_COMMIT_REF_NAME} (${CI_COMMIT_SHA}) with ${CURRENT_STAGING_BRANCH}...`)
+  try {
+    await executeCommand(`git merge --no-ff "${CI_COMMIT_SHA}"`)
+  } catch {
+    const diff = await executeCommand(`git diff`)
```
Comment (BenoitZugmeyer) [inline scripts/staging-ci/merge-into-staging.js:20]: Could you print the caught error too? To be aware if the merge fails with anything other than a conflict we.

File: .gitlab-ci.yml (hunk 4 above)
Comment (amortemousque) [reply .gitlab-ci.yml:387]: Good question, I got inspired by the check-licenses script. I'll try without

File: scripts/staging-ci/merge-into-staging.js (hunk 5 above)
Comment (amortemousque) [reply scripts/staging-ci/merge-into-staging.js:20]: Sure

**Review by BenoitZugmeyer — APPROVED** [review]

**Pull request merged**